pkg/controller/application/app_controller.go

```diff
@@ -0,0 +1,295 @@
+					}
+					crdState := convertShimAppStateToAppCRDState(shimEvent.State)
+					if crdState != "Undefined" {
+						appMgr.updateAppCRDStatus(appCRD, crdState)
+					} else {
+						log.Logger.Error("Invalid status, skip saving it",
+							zap.String("App id", appID))
+					}
+				}
+			}
+		}
+	}
+}
+
+type Controller struct {
+	kubeClient       kubernetes.Interface
+	extensionsClient apiextensionsclient.Interface
+	appClient        applicationclient.Interface
+	informer         cache.SharedIndexInformer
+	lister           applister.ApplicationLister
+}
+
+func NewController(appMgr *AppManager) *Controller {
+	kubeconfig := os.Getenv("KUBECONFIG")
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		log.Logger.Fatal(err.Error())
+	}
+
+	kubeClient := kubernetes.NewForConfigOrDie(config)
+	apiextensionsClient := apiextensionsclient.NewForConfigOrDie(config)
+	appClient := applicationclient.NewForConfigOrDie(config)
+
+	informerFactory := appinformers.NewSharedInformerFactory(appClient, time.Minute*1)
+	informer := informerFactory.Apache().V1alpha1().Applications()
+	informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    appMgr.addApp,
+		UpdateFunc: appMgr.updateApp,
+		DeleteFunc: appMgr.deleteApp,
+	})
+	informerFactory.Start(wait.NeverStop)
+
+	utilruntime.Must(appv1.AddToScheme(appscheme.Scheme))
+	return &Controller{
+		kubeClient:       kubeClient,
+		extensionsClient: apiextensionsClient,
+		appClient:        appClient,
+		informer:         informer.Informer(),
+		lister:           informer.Lister(),
+	}
+}
+
+func (c *Controller) Run() {
+	//TODO: handle the case when the crd is not registered: register it programatically
+	defer utilruntime.HandleCrash()
+
+	log.Logger.Info("Waiting cache to be synced.")
+	// Handle timeout for syncing.
+	timeout := time.NewTimer(time.Second * 30)
+	timeoutCh := make(chan struct{})
+	go func() {
+		<-timeout.C
+		timeoutCh <- struct{}{}
+	}()
+	if ok := cache.WaitForCacheSync(timeoutCh, c.informer.HasSynced); !ok {
+		log.Logger.Error("Timeout expired during waiting for caches to sync. Skip starting the application controller")
+	} else {
+		log.Logger.Info("Starting application controller.")
+	}
+}
+
+/*
+Update the application in the scheduler
+*/
+func (appMgr *AppManager) updateApp(oldObj interface{}, newObj interface{}) {
+	newApp, ok := newObj.(*appv1.Application)
+	if !ok {
+		log.Logger.Error("newObj is not an Application")
+	}
+	oldApp, ok := oldObj.(*appv1.Application)
+	if !ok {
+		log.Logger.Error("oldObj is not an Application")
+	}
+	// No need to update if ResourceVersion is not changed
+	if newApp.ResourceVersion == oldApp.ResourceVersion {
+		log.Logger.Info("No need to update because app object is not modified")
+		return
+	}
+
+	// handle spec change
+	if !equality.Semantic.DeepEqual(oldApp.Spec, newApp.Spec) {
+		//TODO: handle update, discuss what should be updated since only the queue is the common property yet
+	}
```

Comment:
I'd prefer to leave it empty, with a comment

```
// not implement today
```

Reply:
If we will not implement the update, I would rather delete it than leaving it empty